crates/accelerate/src/unitary_synthesis.rs

```diff
@@ -0,0 +1,1094 @@
+            if let (Some(table), Some(matrix)) = (neighbor_table, distance_matrix) {
+                let routing_target = RoutingTargetView {
+                    neighbors: table,
+                    coupling: &table.coupling_graph(),
+                    distance: matrix.as_array(),
+                };
+                // find native gate directions from a (non-bidirectional) coupling map
+                let neighbors0 = &routing_target.neighbors[qubits[0]];
+                let zero_one = neighbors0.contains(&qubits[1]);
+                let neighbors1 = &routing_target.neighbors[qubits[1]];
+                let one_zero = neighbors1.contains(&qubits[0]);
+                match (zero_one, one_zero) {
+                    (true, false) => preferred_direction = Some(true),
+                    (false, true) => preferred_direction = Some(false),
+                    _ => (),
+                }
+            }
+
+            if preferred_direction.is_none() {
+                let mut cost_0_1: f64 = f64::INFINITY;
+                let mut cost_1_0: f64 = f64::INFINITY;
+
+                // Try to find the cost in gate_lengths
+                cost_0_1 = compute_cost(true, &qubits, cost_0_1)?;
+                cost_1_0 = compute_cost(true, &reverse_qubits, cost_1_0)?;
+
+                // If no valid cost was found in gate_lengths, check gate_errors
+                if !(cost_0_1 < f64::INFINITY || cost_1_0 < f64::INFINITY) {
+                    cost_0_1 = compute_cost(false, &qubits, cost_0_1)?;
+                    cost_1_0 = compute_cost(false, &reverse_qubits, cost_1_0)?;
+                }
+
+                if cost_0_1 < cost_1_0 {
+                    preferred_direction = Some(true)
+                } else if cost_1_0 < cost_0_1 {
+                    preferred_direction = Some(false)
+                }
+            }
+        }
+    }
+
+    if natural_direction == Some(true) && preferred_direction.is_none() {
+        return Err(QiskitError::new_err(
+            format!("No preferred direction of gate on qubits {qubits:?} could be determined from coupling map or gate lengths / gate errors.")
+        ));
+    }
+
+    Ok(preferred_direction)
+}
+
+// generic synth function for 2q gates (4x4)
+// used in `run_2q_unitary_synthesis`
+fn synth_su4(
+    py: Python,
+    su4_mat: &Array2<Complex64>,
+    decomposer_2q: &DecomposerElement,
+    preferred_direction: Option<bool>,
+    approximation_degree: Option<f64>,
+) -> PyResult<UnitarySynthesisReturnType> {
+    // double check approximation_degree None
+    let is_approximate = approximation_degree.is_none() || approximation_degree.unwrap() != 1.0;
+    let synth_dag = match &decomposer_2q.decomposer {
+        // the output will be a dag  in the relative basis
+        DecomposerType::XXDecomposer(decomposer) => {
+            let mut kwargs = HashMap::<&str, bool>::new();
+            kwargs.insert("approximate", is_approximate);
+            kwargs.insert("use_dag", true);
```

Comment:
You can do something like this which will save an allocation:
```suggestion
            let kwargs: HashMap<&str, bool> = [
                ("approximate", is_approximate),
                ("use_dag", true)
            ].iter().collect();
```

You can also do `with_capacity(2)` isntead of `new()` for the same effect.

Although in this case it might be better to skip the `HashMap` and build the pydict directly because we never use this rust side you're making an extra object here.